File: tests/conformance/fixtures/go/13-kitchen-sink.go
package main

import sykli "github.com/yairfalse/sykli/sdk/go"

func main() {
	p := sykli.New()

	p.Task("lint").Run("golangci-lint run").
		Inputs("**/*.go").
		Timeout(60).
		SetCriticality("low").
		OnFail("skip")

	p.Task("test").Run("go test ./...").
		After("lint").
		Inputs("**/*.go", "go.mod").
		Env("CGO_ENABLED", "0").
		Retry(2).
		Timeout(300).
		Secrets("CODECOV_TOKEN").
		Matrix("go_version", "1.21", "1.22").
		Service("postgres:15", "db").
		Covers("src/**/*.go").
		Intent("unit tests for all packages").
		SetCriticality("high").
		OnFail("analyze").
		SelectMode("smart")

	p.Task("build").Run("go build -o /out/app").
		After("test").
		Output("binary", "/out/app").
		Provides("binary", "/out/app").
		K8s(sykli.K8sOptions{Memory: "4Gi", CPU: "2"}).
		Target("docker").
		Requires("docker")

	p.Gate("approve-deploy").After("build").
		GateStrategy("env").
		GateTimeout(1800).
		GateMessage("Deploy to production?").
		GateEnvVar("DEPLOY_APPROVED")

	p.Task("deploy").Run("kubectl apply -f k8s/").
		After("approve-deploy").
		Needs("binary").
		SecretFrom("kube_config", sykli.FromFile("/home/.kube/config")).
		SecretFrom("registry_pass", sykli.FromVault("secret/data/registry#password")).
		When("branch:main")

	p.Emit()
}

File: tests/conformance/fixtures/typescript/13-kitchen-sink.ts
import { Pipeline, fromFile, fromVault } from '../../../../sdk/typescript/src/index';

const p = new Pipeline();

p.task('lint').run('golangci-lint run')
  .inputs('**/*.go')
  .timeout(60)
  .setCriticality('low')
  .onFail('skip');

p.task('test').run('go test ./...')
  .after('lint')
  .inputs('**/*.go', 'go.mod')
  .env('CGO_ENABLED', '0')
  .retry(2)
  .timeout(300)
  .secrets('CODECOV_TOKEN')
  .matrix('go_version', '1.21', '1.22')
  .service('postgres:15', 'db')
  .covers('src/**/*.go')
  .intent('unit tests for all packages')
  .setCriticality('high')
  .onFail('analyze')
  .selectMode('smart');

p.task('build').run('go build -o /out/app')
  .after('test')
  .output('binary', '/out/app')
  .provides('binary', '/out/app')
  .k8s({ memory: '4Gi', cpu: '2' })
  .target('docker')
  .requires('docker');

p.gate('approve-deploy').after('build')
  .gateStrategy('env')
  .gateTimeout(1800)
  .gateMessage('Deploy to production?')
  .gateEnvVar('DEPLOY_APPROVED');

p.task('deploy').run('kubectl apply -f k8s/')
  .after('approve-deploy')
  .needs('binary')
  .secretFrom('kube_config', fromFile('/home/.kube/config'))
  .secretFrom('registry_pass', fromVault('secret/data/registry#password'))
  .when('branch:main');

p.emit();
